lighthouse-core/scripts/i18n/collection-util.js

```diff
@@ -0,0 +1,281 @@
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+* Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+*/
+'use strict';
+
+/**
+ * @typedef ICUMessageDefn
+ * @property {string} message
+ * @property {string} [description]
+ * @property {Record<string, ICUPlaceholderDefn>} [placeholders]
+ */
+
+/**
+ * @typedef ICUPlaceholderDefn
+ * @property {string} content
+ * @property {string} [example]
+ */
+
+// @ts-ignore - @types/esprima lacks all of these
+function computeDescription(ast, property, value, startRange) {
+  const endRange = property.range[0];
+  for (const comment of ast.comments || []) {
+    if (comment.range[0] < startRange) continue;
+    if (comment.range[0] > endRange) continue;
+    if (comment.value.includes('@')) {
+      // This is a complex description with description and examples.
+      let description = '';
+      /** @type {Record<string, string>} */
+      const examples = {};
+
+      const r = /@(\w+) ({\w+})?(.*)(\n|$)/g;
+      let resArr;
+      while ((resArr = r.exec(comment.value)) !== null) {
+        const tagName = resArr[1];
+        const placeholder = resArr[2];
+        const message = resArr[3].trim();
+
+        if (tagName === 'description') {
+          description = message;
+        } else if (tagName === 'example') {
+          // Make sure the ICU var exists in string
+          if (!value.includes(placeholder)) {
+            throw Error('Example missing ICU replacement');
+          }
+          examples[placeholder.substring(1, placeholder.length - 1)] = message;
+        }
+      }
+      // Make sure all ICU vars have examples
+      const rMini = /\{(\w+)\}/g;
+      while ((resArr = rMini.exec(value)) !== null) {
+        const varName = resArr[1];
+        if (!examples[varName]) throw Error(`Variable '${varName}' is missing example comment`);
+      }
+      return {description, examples};
+    }
+    return {description: comment.value.replace('*', '').trim()};
+  }
+  return {};
+}
+
+/**
+ * @param {*} message
+ * @param {*} examples
+ * @returns {ICUMessageDefn}
+ */
+function convertMessageToPlaceholders(message, examples = {}) {
+  // console.log(examples);
+  // Basically the same as markdown parsing in dom.js
+  /** @type {Record<string, ICUPlaceholderDefn>} */
+  const placeholders = {};
+
+  // Sanity checks
+  // Number of backticks is even.
+  if ((message.split('`').length - 1) % 2 !== 0) {
+    throw Error(`Open backtick in message "${message}"`);
+  }
+  // Link markdown is not slightly off.
+  if (message.match(/\[.*\] \(.*\)/)) {
```

Comment:
check for any amount/type of space?

worth checking any unmatchedness? I guess this one is less critical since we'd see it in english right away :)

Reply:
Yeah, this is a catch against #9197, I need to add a sister check to this in correct-strings/bake placeholders.

Reply:
`/that's a (?!yes)\w{2} from me dawg/g`